controller/src/test/java/io/pravega/controller/server/eventProcessor/RequestHandlersTest.java

```diff
@@ -547,4 +552,289 @@ private void createStreamInStore(String stream) {
+                new SealStreamTask(streamMetadataTasks, streamTransactionMetadataTasks, streamStore, executor),
+                new DeleteStreamTask(streamMetadataTasks, streamStore, bucketStore, executor),
+                new TruncateStreamTask(streamMetadataTasks, streamStore, executor),
+                streamStore,
+                executor);
+        String fairness = "fairness";
+        streamStore.createScope(fairness).join();
+        streamMetadataTasks.createStream(fairness, fairness, StreamConfiguration.builder().scalingPolicy(ScalingPolicy.fixed(1)).build(),
+                System.currentTimeMillis()).join();
+
+        // 1. set segment helper mock to throw exception
+        doAnswer(x -> Futures.failedFuture(new RuntimeException()))
+                .when(segmentHelper).truncateSegment(anyString(), anyString(), anyLong(), anyLong(), any(), any(), anyString(), anyLong());
+        
+        // 2. start process --> this should fail with a retryable exception while talking to segment store!
+        streamStore.startTruncation(fairness, fairness, Collections.singletonMap(0L, 0L), null, executor).join();
+        streamStore.setState(fairness, fairness, State.TRUNCATING, null, executor).join();
+        assertEquals(State.TRUNCATING, streamStore.getState(fairness, fairness, true, null, executor).join());
+        
+        TruncateStreamEvent event = new TruncateStreamEvent(fairness, fairness, 0L);
+        AssertExtensions.assertFutureThrows("", streamRequestHandler.process(event),
+                e -> Exceptions.unwrap(e) instanceof RuntimeException);
+
+        verify(segmentHelper, atLeastOnce()).truncateSegment(anyString(), anyString(), anyLong(), anyLong(), any(), any(), anyString(), anyLong());
+        
+        // 3. set waiting processor to "random name"
+        streamStore.createWaitingRequestIfAbsent(fairness, fairness, "myProcessor", null, executor).join();
+        
+        // 4. reset segment helper to return success
+        doAnswer(x -> CompletableFuture.completedFuture(null))
+                .when(segmentHelper).truncateSegment(anyString(), anyString(), anyLong(), anyLong(), any(), any(), anyString(), anyLong());
+        
+        // 5. process again. it should succeed while ignoring waiting processor
+        streamRequestHandler.process(event).join();
+        assertEquals(State.ACTIVE, streamStore.getState(fairness, fairness, true, null, executor).join());
+        
+        // 6. run a new update. it should fail because of waiting processor.
+        TruncateStreamEvent event2 = new TruncateStreamEvent(fairness, fairness, 0L);
+        AssertExtensions.assertFutureThrows("", streamRequestHandler.process(event2),
+                e -> Exceptions.unwrap(e) instanceof StoreException.OperationNotAllowedException);
+        streamStore.deleteWaitingRequestConditionally(fairness, fairness, "myProcessor", null, executor).join();
+    }
+    
+    @Test
+    public void testCommitTxnIgnoreFairness() {
```

Comment:
If I'm not wrong, we found the problem in a test that, in addition to failovers, we interleaved scale stream and commit transaction events (`ReadTxnWriteAndAutoScaleWithFailoverTest`). Does it worth it to create a test that verifies that scale and commit events do not get deadlocked?

Reply:
This test is doing exactly that but in slightly more generic fashion. we set waiting processor on scale and commit transcation while they have been started.